allennlp/data/dataset_readers/huggingface_datasets_reader.py

```diff
@@ -0,0 +1,190 @@
+from allennlp.data.instance import Instance
+from datasets import load_dataset, Dataset
+from datasets.features import ClassLabel, Sequence, Translation, TranslationVariableLanguages
+from datasets.features import Value
+
+
+class HuggingfaceDatasetSplitReader(DatasetReader):
+    """
+    This reader implementation wraps the huggingface datasets package
+    to utilize it's dataset management functionality and load the information in AllenNLP friendly formats
+    Note: Reader works w.r.t to only one split of the dataset,
+    i.e. you would need to create separate reader for separate splits
+
+    Following dataset and configurations have been verified and work with this reader
+
+            Dataset                       Dataset Configuration
+            `xnli`                        `ar`
+            `xnli`                        `en`
+            `xnli`                        `de`
+            `xnli`                        `all_languages`
+            `glue`                        `cola`
+            `glue`                        `mrpc`
+            `glue`                        `sst2`
+            `glue`                        `qqp`
+            `glue`                        `mnli`
+            `glue`                        `mnli_matched`
+            `universal_dependencies`      `en_lines`
+            `universal_dependencies`      `ko_kaist`
+            `universal_dependencies`      `af_afribooms`
+            `afrikaans_ner_corpus`        `NA`
+            `swahili`                     `NA`
+            `conll2003`                   `NA`
+            `dbpedia_14`                  `NA`
+            `trec`                        `NA`
+            `emotion`                     `NA`
+    """
+
+    def __init__(
+        self,
+        max_instances: Optional[int] = None,
+        manual_distributed_sharding: bool = False,
+        manual_multiprocess_sharding: bool = False,
+        serialization_dir: Optional[str] = None,
+        dataset_name: str = None,
+        split: str = "train",
+        config_name: Optional[str] = None,
+    ) -> None:
+        super().__init__(
+            max_instances,
+            manual_distributed_sharding,
+            manual_multiprocess_sharding,
+            serialization_dir,
+        )
+
+        # It would be cleaner to create a separate reader object for different dataset
+        self.dataset: Dataset = None
+        self.dataset_name = dataset_name
+        self.config_name = config_name
+        self.index = -1
+
+        if config_name:
+            self.dataset = load_dataset(self.dataset_name, self.config_name, split=split)
+        else:
+            self.dataset = load_dataset(self.dataset_name, split=split)
+
+    def _read(self, file_path) -> Iterable[Instance]:
+        """
+        Reads the dataset and converts the entry to AllenNLP friendly instance
+        """
+        if self.dataset:
+            for entry in self.dataset:
```

Comment:
To make this efficient with multi-process data loading, consider using `self.shard_iterable()` here like:

```suggestion
            for entry in self.shard_iterable(self.dataset):
```

But I know the `datasets` has a `.select()` method, which you could potentially utilize like:

```suggestion
            for entry in self.dataset.select(self.shard_iterable(range(self.dataset.num_rows))):
```

That might be more efficient, I'm not sure.

Reply:
I see `Dataset.select()` doc indicates it makes into a separate dataset. I will go without `select()` for now. Will keep the conv open to re-visit and confirm the better choice before we merge.

Reply:
I have used self.shard_iterable. 
Let me add a TODO for checking the select() support

Reply:
Have added a TODO for this.